tests/file_storage/test_inmemory_storage.py

```diff
@@ -49,6 +50,14 @@ def test_file_saving(self):
         self.storage.save("file.dat", ContentFile(b"test"))
         self.assertEqual(self.storage.open("file.dat", "rb").read(), b"test")
 
+    def test_file_save_without_name(self):
+        """
+        File storage extracts the filename from the content object if no
+        name is given explicitly.
+        """
+        self.storage.save(None, InMemoryFileNode("test", name="file.txt"))
```

Comment:
InMemoryFileNode is an undocumented class and testing it directly doesn't feel like we're capturing the issue from the ticket

I can't replicate the same issue from the ticket but I do get an error with
```diff
diff --git a/tests/file_storage/tests.py b/tests/file_storage/tests.py
index c7ca873521..9e9722b0fd 100644
--- a/tests/file_storage/tests.py
+++ b/tests/file_storage/tests.py
@@ -966,6 +966,22 @@ class FileFieldStorageTests(TestCase):
         self.assertEqual(obj.empty.read(), b"more content")
         obj.empty.close()
 
+    @override_settings(
+        STORAGES={
+            DEFAULT_STORAGE_ALIAS: {
+                "BACKEND": "django.core.files.storage.InMemoryStorage",
+            },
+        }
+    )
+    def test_in_memory_storage_without_name(self):
+        f = ContentFile("content", "file.txt")
+        obj = Storage()
+        obj.storage_callable_default.save(None, f)
+        self.assertEqual(
+            repr(obj.storage_callable_default),
+            "<FieldFile: storage_callable_default/file.txt>"
+        )
+
     def test_pathlib_upload_to(self):
```
Which makes me think we should be raising an error when `name` is none or try to derive it from the file

Reply:
Thank you for the review :open_hands: 

> InMemoryFileNode is an undocumented class and testing it directly doesn't feel like we're capturing the issue from the ticket

My decision to test `InMemoryFileNode` directly instead of using `FileField` as described on ticket was to make the test more unitary.

The issue in my opinion is not on `FileField` or any other class that may use `InMemoryFileNode`, but the problem lies within `InMemoryFileNode`.

Although `InMemoryFileNode`  is undocumented, on places such as `Storage.save` it seems to me that there is an expectation of all `File` subclasses to have a name attribute.
```
    def save(self, name, content, max_length=None):
        """
        Save new content to the file specified by name. The content should be
        a proper File object or any Python file-like object, ready to be read
        from the beginning.
        """
        # Get the proper name for the file, as it will actually be saved.
        if name is None:
            name = content.name
```

Nonetheless, I don't think it's a necessarily a better or worse decision than testing `FileField` directly. If you think a `FileField` would be more accurate, I can add the `FileField` test instead. However, so far I couldn't replicate the case mentioned by David on the ticket either.



> Which makes me think we should be raising an error when `name` is none or try to derive it from the file

**About deriving the file name from the file when it's None:**
We're already trying to derive it from the file at `File.__init__`:
```
    def __init__(self, file, name=None):
        self.file = file
        if name is None:
            name = getattr(file, "name", None)
        self.name = name
        if hasattr(file, "mode"):
            self.mode = file.mode
```
`File` subclasses re-use this logic by calling `File` constructor from their own constructors, except for `InMemoryFileNode` that receives the `name` parameter but does nothing with it and does not pass it up to its parent class.

**About raising an exception when it's None:**
I think this is also something we should do.
However, I think we shouldn't raise an exception at `InMemoryFileNode`, since we don't do it for any other `File` subclass. Also, we can't know in advance whether the user intend to set a file name at `Storage.save` later on.

Still, there is the potential inconvenience of any kind of `File` having its name set to `None`, and then trying to save it into a `Storage` by passing once again `None` as name. For example, I added the following tests and they indeed fail for me, regardless the type of `File` or `Storage` used:

```
class FileStorageTests(SimpleTestCase):
    storage_class = FileSystemStorage
    
    def test_fail_if_no_name_is_given(self):
        file = ContentFile("custom contents")
        with self.assertRaises(ValueError):
            self.storage.save(None, file)

    def test_fail_if_both_file_names_are_empty(self):
        file = ContentFile("custom contents", name="")
        with self.assertRaises(ValueError):
            self.storage.save("", file)
```

I think we can indeed raise a more clear exception on such cases.
I addressed this issue on a [new commit](https://github.com/django/django/pull/18453/commits/67faa5bff1c9ffe4c3c330d77b6caa7b91b2576a).

Reply:
Quick update:
Since the user who opened the ticket gave us a proper way to reproduce the issue from `Filefield` in [his latest comment](https://code.djangoproject.com/ticket/35658#comment:13), I added it to a test case on [another commit](https://github.com/django/django/pull/18453/commits/4953efa82967982e690339d7fde0033208c59704) as requested